### PR TITLE
Add support for fetching entity titles to display as breadcrumbs

### LINKED ISSUE #2860: Showing titles instead of IDs in the navigation bar
### Is your feature request related to a problem?
When looking through different courses, exams and exercises the navigation link bar below the Artemis sign is not as useful as it could be, because only the IDs are shown and not the titles.

artemis_navigation

### Describe the solution you'd like
Would it be possible to replace the IDs (e.g. 95 in the screenshot) with the corresponding course, exam and exercise titles? This way I would always know where I am without looking it up again or remembering IDs.

### PR DESCRIPTION
**Pull request by FrankeLukas**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).
- [x] Client: I documented the TypeScript code using JSDoc style.

### Motivation and Context

For non-admins and non-instructors IDs would be displayed instead of titles in several cases, since normal users and teach assisstants didn't have the permissions to fetch the entity to read the title from.
Closes #2860.

### Description

We introduce new methods to fetch only the titles, such that no important data is leaked when an arbitrary user executes the query.

### Steps for Testing

Test that breadcrumbs display the correct names instead of IDs for:
- Courses
- Lectures
- Exams
- Exercises
- Exercise Hints
- Apollon Diagrams

For example, on develop tutors cannot fetch the entire exam, hence an error is displayed:
grafik
And the breadcrumbs look like this (notice the ID):
grafik

With this PR there is no error and the breadcrumbs look like the following:
grafik

### Test Coverage

I adapted the client side navbar test which uses the new client side methods and introduced server-side tests for every new REST endpoint. => 100%

### CODE REVIEW

**Review by daniels98it — APPROVED** [review]
Tested in testing session..
Works good :+1:

**Review by nicolasruscher — DISMISSED** [review]
tested in session, works!

**Review by philippbauch — APPROVED** [review]
Tested in testing session: seems to work fine 👍🏻

**Review by ivohashamov — DISMISSED** [review]
Tested on ts1, works as intended

**Review by FuchsDominik — APPROVED** [review]
Looks good, tested the routes and breadcrumbs seem to look as advertised
Screenshot 2021-03-23 at 17 09 18

File: src/main/java/de/tum/in/www1/artemis/repository/LectureRepository.java [hunk 1]
```diff
@@ -44,4 +44,16 @@
...
+     *
+     * @param lectureId the id of the lecture
+     * @return the name/title of the lecture
+     */
+    @Query("""
+            SELECT l.title
+            FROM Lecture l
+            WHERE l.id = :lectureId
+            """)
+    String getLectureTitle(Long lectureId);
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/repository/LectureRepository.java:58]: what happens if the id cannot be found? does it return null? do we need to handle this case here?

File: src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java [hunk 2]
```diff
@@ -1120,6 +1121,19 @@ public void fetchParticipationsWithSubmissionsAndResultsForCourses(List<Course>
...
+     * GET /courses/:courseId/get-title : Returns the title of the course with the given id
+     *
+     * @param courseId the id of the course
+     * @return the name/title of the course
+     */
+    @GetMapping(value = "/courses/{courseId}/get-title")
+    @PreAuthorize("hasAnyRole('USER', 'TA', 'INSTRUCTOR', 'ADMIN')")
+    @ResponseBody
+    public StringDTO getCourseTitle(@PathVariable Long courseId) {
+        return new StringDTO(courseRepository.getCourseTitle(courseId));
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java:1134]: what happens if the return value is null? how is this handled?

File: src/main/java/de/tum/in/www1/artemis/repository/LectureRepository.java (hunk 1 above)
Comment (FrankeLukas) [reply src/main/java/de/tum/in/www1/artemis/repository/LectureRepository.java:58]: The method will return `null`. We don't need to handle that case. (See comment below.)

File: src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java (hunk 2 above)
Comment (FrankeLukas) [reply src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java:1134]: If the return value of the repository call is `null` that same value is wrapped and sent to the client, which falls back to the ID if `undefined` or `null` is returned. I'll check what `ResponseEntity<String>` does in this case.

**Comment by FrankeLukas** [comment]
Updated and resolved the merge conflicts, we now explicitly return a 404 Not Found if an entity with the given ID cannot be found:
grafik

**Review by nicolasruscher — APPROVED** [review]
Re-approving

**Review by SimonEntholzer — APPROVED** [review]
Tested on TS2 now tutors can see the title without the 403 error. :+1: , code looks good too. A lot of tests which is nice :D

**Review by ivohashamov — APPROVED** [review]
Tested on ts2, exam titles appear as expected and no errors get thrown. Looks good! Reapprove.

**Pull request merged**